docs/guides/deno/byo-deep-research.mdx

```diff
@@ -0,0 +1,738 @@
+3. Be careful not to generate follow-up queries that are not related to the topics.
+4. Generate a follow-up question that would help expand your understanding
+</GOAL>
+
+<REQUIREMENTS>
+Ensure the follow-up question is self-contained and includes necessary context for web search.
+</REQUIREMENTS>
+
+Provide only the follow-up query in your response, if there are no follow-up queries, respond with nothing.
+`
+```
+
+These prompts work together to create a research system that:
+
+- Generate search queries from topics
+- Find relevant content using the duckduckgo search API
+- Cleans and converts content to simple markdown
+- Summarizes findings
+- Attempts to identify knowledge gaps
+- Continues research if required
+
+### 3. Topic-based Research Pipeline
+
+The system uses a topic-based approach to conduct research, with different message types for each stage of the process:
+
+```typescript title:services/api.ts
+type TopicMessageTypes = 'create_query' | 'query' | 'reflect' | 'summarize'
+
+interface ResearchTopicMessage<T extends TopicMessageTypes> {
+  type: T
+  topics: string[]
+  summaries: string[]
+  remainingIterations: number
+}
+
+interface CreateQueryTopicMessage extends ResearchTopicMessage<'create_query'> {
+  type: 'create_query'
+  date: string
+  originalTopic: string
+}
+
+interface PerformQueryTopicMessage extends ResearchTopicMessage<'query'> {
+  type: 'query'
+  query: {
+    query: string
+    rationale: string
+  }
+}
+
+interface SummarizeTopicMessage extends ResearchTopicMessage<'summarize'> {
+  type: 'summarize'
+  content: string
+}
+
+interface ReflectTopicMessage extends ResearchTopicMessage<'reflect'> {
+  type: 'reflect'
+  content: string
+}
+```
+
+### 4. Content Processing
+
+The system includes HTML cleaning and markdown conversion:
+
+```typescript title:services/api.ts
+import { load as cheerioLoad } from 'cheerio'
+import { TurndownService } from 'turndown'
+
+// Initialize TurndownService for HTML to Markdown conversion
+const turndownService = new TurndownService({
+  headingStyle: 'atx',
+  codeBlockStyle: 'fenced',
+})
+
+function cleanHtml(html: string): string {
+  const $ = cheerioLoad(html)
+
+  // Remove script and style tags
+  $('script, style, noscript, iframe, embed, object').remove()
+
+  // Remove navigation elements
+  $('nav, header, footer, aside, .nav, .navigation, .menu, .sidebar').remove()
+
+  // Remove ads and social media elements
+  $('.ad, .ads, .advertisement, .social, .share, .comments').remove()
+
+  // Remove empty elements
+  $('*').each(function () {
+    if ($(this).text().trim() === '') {
+      $(this).remove()
+    }
+  })
+
+  // Get the main content
+  let content = $('article, main, .article, .content, .post, .entry')
+  if (content.length === 0) {
+    content = $('body')
+  }
+
+  return content.html() || ''
+}
+```
+
+### 5. The Nitric API Implementation
+
+The main API implementation ties everything together:
```

Comment:
This is close to a 300 line block of code. I think a bit of explanation beforehand, or splitting it into more digestable chunks will be much more readable as a guide. You can add a note linking to the file on GitHub if someone wants to copy the entire chunk and/or read it from there.

Reply:
I think this is a good suggestion. In the interest of getting this out I might leave this as is for now and think of a more effective method we can you to help decompose this section (possibly taking advantage of codehike?)